Fix/check bookmark metoo

yesjjin99 opened the pull request:
- 스터디 디테일 조회 시 북마크 여부, 참가 신청 여부 함께 반환 (#65)
- 댓글 리스트 조회 시 나도 궁금해요 등록 여부 함께 반환

Jeyoung-Park commented on src/routes/study/comment/comment.controller.ts:
```diff
@@ -1,21 +1,83 @@
+        accessToken,
+        process.env.SIGNUP_TOKEN_SECRET as string
+      ) as { id: string };
+      req.user = { id: decoded.id };
```
40~44번쨰 줄과 코드가 중복되는 부분으로 보이는데 중복을 줄여도 좋을 것 같습니다. 중요도가 높진 않아서 굳이 반영은 안 해주셔도 괜찮을 것 같아요 👍

Jeyoung-Park replied:
엇 다시 보니 그 아래 부분까지 겹치는 부분이 있는 것 같아서 이 부분 통일하면 좋을 것 같네요!

Jeyoung-Park commented:
@yesjjin99 오오 북마크 여부만 반환해 주시는 줄 알았는데, 나도 궁금해요와 참가 신청 여부까지 같이 작업해주셨군요..!! 고생 많으셨을텐데 감사합니다 😄

rudy3091 commented on src/routes/study/comment/comment.controller.ts:
```diff
@@ -1,21 +1,83 @@
+    /* checkToken */
+    const { accessToken, refreshToken } = req.cookies;
+
+    const logout_result: any[] = [];
```
개인적으로 any를 쓰는건 이후 부채를 너무 많이 만들어내 지양하는게 좋다고 생각합니다
이후에 사용되는 걸 보면 `Array<Comment & { metoo: boolean }>` 정도로  정리해볼 수 있지 않을까요?
그나저나 얘는 갑자기 왜 스네이크 케이스인가요??ㅋㅋㅋㅋ

yesjjin99 replied:
급하게 하니까 코드가 더 엉망이네요ㅎ
카멜 케이스로 통일했고, 마찬가지로 타입도 변경했습니다!
꼼꼼한 리뷰 너무 감사합니다ㅠㅠ

rudy3091 commented on src/routes/study/comment/comment.controller.ts:
```diff
@@ -1,21 +1,83 @@
+    }
+
+    if (!accessToken && refreshToken) {
+      await refresh(req, res);
```
refresh 는 내부적으로 `res.json()` 메소드를 호출하는데, `res.json()` 메소드는 또 내부적으로 `res.end()` 메소드를 호출해 응답을 클라이언트로 전송완료해버립니다
지금 이 로직을 봤을땐 `cannot set status code after they are sent to client` 뭐시기 에러가 발생할 것 같은데 체크해보셨나요??

yesjjin99 replied:
해당 이슈 때문에 refresh 미들웨어 호출하고 나서 제대로 액세스토큰이 재발급되었는지 확인한 후, 실패했으면 return 처리해주고 성공한 경우 해당 액세스토큰을 받아오도록 수정하기는 했습니다
일단 다시 체크해봐야 할 것 같기는 하네요

rudy3091 commented on src/routes/study/comment/comment.controller.ts:
```diff
@@ -1,21 +1,83 @@
+            };
+          }
+        }
+        return res.status(200).json(result);
+      } catch (e) {
+        return res.status(200).json(logout_result);
```
에러가 발생했을때도 200 코드로 응답해주나요?? 다음 회의때 로직 설명을 조금 들어보고 싶습니다 🙏

yesjjin99 replied:
넵, 해당 로직에 대해서는 회의 때 간단하게 이야기해볼 생각입니다
간단하게 이야기하자면 댓글 리스트 조회가 로그인 여부 상관없이 가능한 api라서 로그아웃 상태일 때는 metoo flag를 전부 false로 응답해주는 방식으로 작업해서 401을 응답해주는 대신 200을 응답하도록 작성했습니다

rudy3091 commented on src/routes/study/comment/comment.controller.ts:
```diff
@@ -57,25 +119,21 @@ const createComment = async (req: Request, res: Response) => {
       } else {
-        const notiTitle = '새로운 문의글';
-        const notiAbout = '새 문의글이 달렸어요.';
-        await createStudyNoti(
-          studyid,
-          study.HOST_ID,
-          notiTitle,
-          notiAbout,
-          102
-        );
+        await createStudyNoti({
+          id: studyid,
+          userId: study.HOST_ID,
+          title: '새로운 문의글',
+          about: '새 문의글이 달렸어요.',
+          type: NotiTypeEnum.NEW_COMMENT,
```
뭔가 구조가 반복되는 것 같은데 enum 으로 따로 빼서 사용해도 좋을 것 같네요
배포 이후에 대대적인 리팩토링이 필요할것 같습니다 🤔

yesjjin99 replied:
동의합니다ㅠ 지금 해주기에는 조금 무리가 있을 것 같아서 일단 이대로 두고 배포 후에 전부 리팩토링을 해줄 필요가 있을 것 같아요

rudy3091 commented on src/services/study/index.ts:
```diff
@@ -179,31 +179,34 @@ const createStudy = async (studyDTO: studyDTO, user: UserProfile) => {
-              const notiAbout = `모집이 종료되었어요. 스터디를 응원합니다!`;
-              await createStudyNoti(
-                studyId,
-                member?.user.id,
-                notiTitle,
-                notiAbout,
-                107
-              );
+              await createStudyNoti({
+                id: studyId,
+                userId: member.userId,
+                title: '모집 종료',
+                about: `모집이 종료되었어요. 스터디를 응원합니다!`,
+                type: NotiTypeEnum.CLOSED,
+              });
             }
```
지금 든 생각인데, 알림을 생성하는 로직이 서비스 레이어가 아니라 컨트롤러 레이어에 있어야 한다는 생각이 드네요..
실제로 어떤 로직은 컨트롤러 레이어에 위치하고 어떤 로직은 서비스 레이어에 존재하고 있네요
study service 레이어는 Study 엔티티에 관련한 데이터베이스 연산이 들어가는 곳인데 notification service 에 디펜던시가 생기는건 좋지 않은 패턴으로 보입니다
이부분도 지금 고치기엔 너무 공수가 많이 들어갈 것 같고.. 배포 이후 리팩토링 때 고치는 거로 해도 좋을것 같아요

yesjjin99 replied:
스터디 생성을 제외한 나머지 케이스들은 컨트롤러 레이어에서 알림을 생성하고 있어서 저도 고민을 했었던 부분입니다...!
그냥 지금 작업하는 김에 스케줄링 로직과 함께 알림 생성 로직까지 전부 컨트롤러에서 처리하도록 수정했습니다
이 이상의 공수는 말씀해주신 것처럼 지금 단계에서는 힘들 것 같고, 배포 이후에 리팩토링 들어가면서 함께 작업하면 좋을 것 같습니다!

rudy3091 reviewed (commented):
고생하셨습니다! 코멘트 남겼습니다 :)

yesjjin99 committed:
Fix: Add bookmarked, metoo, applied flag

댓글 조회 시 나도 궁금해요 여부 함께 반환

Add bookmarked flag to study detail API

Refactor notification

Add applied flag to study detail API

yesjjin99 committed:
Merge branch 'develop' into fix/check-bookmark-metoo
